Artix Linux Commits

Dudemanguy opened the pull request:
No idea if upstream is alive or not, but I guess I'll leave this set of commits here for visibility's sake. Each one is pretty self-explanatory I think. This is what what I've fixed and rewritten for Artix Linux.
https://gitea.artixlinux.org/artix/pam_rundir

Dudemanguy committed:
fix group permissions of the user folders

When making a directory, the user was correctly given ownership but the
group was still root. In elogind (and likely system-logind), these
folders are owned by user:user. Fix this by also setting the egid to
the user's group.

Dudemanguy committed:
configure: default to /run/user

elogind (and presumably system-logind) uses /run/user. Copy them for
consistency.

Dudemanguy committed:
cosmetic clean up

Alphabetize the headers and remove the extra indention in
pam_sm_open_session function.

Dudemanguy committed:
simplify and replace memmove/memcpy with snprintf

The old way used memmove/memcpy when altering paths since it was writing
to .$uid and $uid. We don't need to worry about that case so the path
handling can be made simplier by just using snprintf instead. It's
probably better practice in general for combining paths.

Dudemanguy committed:
use utmp instead of writing counters to files

The original implementation worked by updating a counter in a file
(/run/user/.$uid) anytime a user logged in. If the counter goes to 0,
remove the directory. This works fine, but is a bit ugly and has some
complicated logic/pitfalls. Instead, we can just use utmp to keep track
of logins. By looping through the LOGIN_PROCESS and USER_PROCESS, we can
determine whether or not a user has completely logged out of all
sessions (and thus remove the /run/user/$uid directory) without needing
to write any of the .$uid files. This simplifies the code and makes it
cleaner.

jirutka commented:
FYI: I’ve backported 7796f6c and 63bd5b1 to the package [pam-rundir](https://pkgs.alpinelinux.org/packages?name=pam-rundir&branch=edge) in Alpine Linux.

> use utmp instead of writing counters to files (4a15d66)

This one is problematic for Alpine Linux and other musl-based distributions, because musl does not implement utmp. We’re currently working on wider integration of [utmps](https://pkgs.alpinelinux.org/packages?name=utmps&branch=edge) that provides utmp/wtmp support for musl, but the services needed to make it work will be optional.

Thus if this MR is merged, I’d appreciate adding a compile flag to switch between the current implementation with a counter in a file and the utmp-based implementation. Also, please note that utmp as implemented in glibc is not secure:

> utmp functionality is difficult to implement in a secure way; in particular, it is impossible to implement without either running a daemon or allowing arbitrary programs to tamper with user accounting. – https://skarnet.org/software/utmps/

Dudemanguy commented:
I could use utmpx.h instead that's helpful. I didn't realize musl didn't implement utmp. As for security concerns, it's not any different than the old way. If some program tampered with /run/user/.1000, things can go wrong. I guess that wouldn't be as big of a target though.

jirutka commented:
> I could use utmpx.h instead that's helpful.

Well, you should definitely use `utmpx.h` instead of `utmp.h` – the latter is deprecated and actually non-standard.

musl provides utmp(x) stub, so it could be built on musl, but the utmp(x) functionality will do nothing.

Dudemanguy commented:
>the latter is deprecated

Wasn't aware of that. I'll do that. Thanks!

Dudemanguy committed:
use utmpx instead of utmp

The latter is deprecated actually.

Dudemanguy committed:
utmp: update state for any user login

Previously, it was thought that only the user in question needed to be
taken into account for utmp, but this is not correct. Sometimes, a
USER_PROCESS would be recorded for one user in a tty then a short while
later is it followed by another USER_PROCESS for another user in the
same tty. No LOGIN_PROCESS with a null user is neccesarily recorded.
Without taking this into account, there was a bug in the logic and a
directory could mistakenly be deleted (or not deleted).

jirutka commented on pam_rundir.c:
```diff
@@ -404,8 +404,8 @@ pam_sm_open_session (pam_handle_t *pamh, int flags, int argc, const char **argv)
-        /* set euid so if we do create the dir, it is own by the user */
-        if (seteuid (pw->pw_uid) < 0)
+        /* set euid and egid so if we do create the dir, it is owned by the user */
+        if (seteuid (pw->pw_uid) < 0 || setegid (pw->pw_gid) < 0)
```
pam_rundir creates `/run/user` owned by _root:root_ and mode _775_. If the target user is not member of group _root_, pam_rundir cannot create a directory inside it after switching uid/gid.

jirutka replied:
Aha, no, it’s not an issue on branch `next`. The problem is that I backported this change to the latest release without 5ae2ee1a3411e973ebe56421850b7e08c23d6319. :/
